cql3: expr: add unit tests for prepare_expression

cvybhu opened the pull request:
Adds unit tests for the function `expr::prepare_expression`.

Three minor bugs were found by these tests, both fixed in this PR.
1. When preparing a map, the type for tuple constructor was taken from an unprepared tuple, which has `nullptr` as its type.
2. Preparing an empty nonfrozen list or set resulted in `null`, but preparing a map didn't. Fixed this inconsistency.
3. Preparing a `bind_variable` with `nullptr` receiver was allowed. The `bind_variable` ended up with a `nullptr` type, which is incorrect. Changed it to throw an exception,

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2941/

avikivity commented on test/lib/expr_test_utils.cc:
```diff
@@ -389,6 +389,121 @@ std::pair<evaluation_inputs, std::unique_ptr<evaluation_inputs_data>> make_evalu
 
+mock_database_impl::mock_database_impl(schema_ptr table_schema) {
+    _table_schema = table_schema;
+    _keyspace_metadata = data_dictionary::keyspace_metadata::new_keyspace(
```
Please use the regular constructor member initializer list style, it makes it easier to add things later if they don't have default constructors.

avikivity commented on test/lib/expr_test_utils.cc:
```diff
@@ -389,6 +389,121 @@ std::pair<evaluation_inputs, std::unique_ptr<evaluation_inputs_data>> make_evalu
+}
+
+const secondary_index::secondary_index_manager& mock_database_impl::get_index_manager(data_dictionary::table t) const {
+    throw std::bad_function_call();
```
Note: we will have to move the index stuff into data_dictionary later. Out of scope for now.

avikivity commented on test/lib/expr_test_utils.hh:
```diff
@@ -98,6 +98,57 @@ std::pair<evaluation_inputs, std::unique_ptr<evaluation_inputs_data>> make_evalu
     const std::vector<raw_value>& bind_marker_values = {});
+
+// A mock implementation of data_dictionary::database, used in tests
+class mock_database_impl : public data_dictionary::impl {
```
You can move this to .cc too. Nobody needs the interface.

avikivity commented on test/lib/expr_test_utils.hh:
```diff
@@ -98,6 +98,57 @@ std::pair<evaluation_inputs, std::unique_ptr<evaluation_inputs_data>> make_evalu
+
+    db::config _config;
+
+   public:
```
bad indent.

cvybhu replied:
It's a new file so I took the chance to keep it compliant with `clang-format`.
Adding `AccessModifierOffset: -4` fixed this.

avikivity commented on test/lib/expr_test_utils.hh:
```diff
@@ -98,6 +98,57 @@ std::pair<evaluation_inputs, std::unique_ptr<evaluation_inputs_data>> make_evalu
+    db::config _config;
+
+   public:
+    mock_database_impl(schema_ptr table_schema);
```
explicit

avikivity commented on test/lib/expr_test_utils.hh:
```diff
@@ -98,6 +98,57 @@ std::pair<evaluation_inputs, std::unique_ptr<evaluation_inputs_data>> make_evalu
+};
+
+// Creates a mock implementation of data_dictionary::database, useful in tests.
+std::pair<data_dictionary::database, std::unique_ptr<mock_database_impl>> make_data_dictionary_database(
```
If you change this to return std::pair<..., std::unique_ptr<data_dictionary::impl>> it will still work, and allow you to move the definition out-of-line.

avikivity commented on test/boost/expr_test.cc:
```diff
@@ -1303,4 +1303,56 @@ BOOST_AUTO_TEST_CASE(evaluate_bind_variable_validates_empty_value_in_maps_recurs
-}
\ No newline at end of file
+}
+
```
> It's interesting that prepare_expression
for column identifiers doesn't require a receiver.
I hope this won't break validation in the future.

prepare_expression has four modes:
 - with a receiver 
     - with the prepared expression having a known type
        check that the types are compatible
     - with the prepared expression not having a known type (like bind variables)
        infer the prepared type from the reciever
 - without a receiver
     - with the prepared expression having a known type
        accept the prepared expression type
     - with the prepared expression not having a known type (like bind variables)
        reject the expression

For any given expression just two are applicable, and we should test both of them. What you have now is okay to start with.

avikivity commented on test/boost/expr_test.cc:
```diff
@@ -1366,3 +1366,92 @@ BOOST_AUTO_TEST_CASE(prepare_column_value) {
+
+    expression expected = subscript{.val = column_value(table_schema->get_column_definition("r")),
+                                    .sub = make_int_const(123),
+                                    .type = boolean_type};
```
It's better to use some other uncommon type, like float. Expressions can generate booleans, so if the test succeeds you don't know if it's because it arrived at the type correctly, or got it from some boolean expression accidentally. It's not a problem here, but note it in the future.

avikivity commented on test/boost/expr_test.cc:
```diff
@@ -1545,3 +1545,21 @@ BOOST_AUTO_TEST_CASE(prepare_null_no_type_fails) {
+
+    ::lw_shared_ptr<column_specification> receiver = make_receiver(int32_type);
+
+    expression prepared = prepare_expression(bind_var, db, "test_ks", table_schema.get(), receiver);
```
Should also test the second case (no receiver -> failed type inference)

cvybhu replied:
Added, although it turned out that we don't check if the receiver for bind_variable exists x.x
Fixed the bug in the previous commit.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -246,6 +246,12 @@ map_prepare_expression(const collection_constructor& c, data_dictionary::databas
+        on_internal_error(expr_logger,
+                          format("map_prepare_expression bad non-map receiver type: {}", receiver->type->name()));
+    }
+    data_type map_element_tuple_type = tuple_type_impl::get_instance({map_type->get_keys_type(), map_type->get_values_type()});
```
Please adjust the commitlog to explain the bug and the fix.

cvybhu replied:
Added, forgot to write it.

avikivity commented:
Very good.

cvybhu commented on test/boost/expr_test.cc:
```diff
@@ -1660,3 +1660,73 @@ BOOST_AUTO_TEST_CASE(prepare_untyped_constant_bad_int) {
+
+    BOOST_REQUIRE_THROW(prepare_expression(tup, db, "test_ks", table_schema.get(), nullptr),
+                        exceptions::invalid_request_exception);
+}
```
Another test prepares a `tuple_constructor` of columns without a receiver and it succeeds.

cvybhu commented:
v2:
* Moved mock_database_impl to the `.cc` file
* Add `explicit` to mock_database_impl constructor
* Use constructor member initializer list style in mock_database_impl constructor
* Added a test that prepares a bind_variable without a receiver
* Fixed a bug which allowed to prepare a bind_variable without a receiver, trying to do this now throws an exception
* Added commit message in the commit that fixes preparing map tuple_constructors

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3053/
